sw/device/lib/dif/dif_hmac.c

```diff
@@ -0,0 +1,343 @@
+      break;
+    case kDifHmacEndiannessLittle:
+      mmio_region_nonatomic_clear_bit32(hmac->base_addr, HMAC_CFG_REG_OFFSET,
+                                        HMAC_CFG_ENDIAN_SWAP);
+      break;
+    default:
+      return kDifHmacError;
+  }
+
+  return kDifHmacOk;
+}
+
+dif_hmac_result_t dif_hmac_digest_endianness_set(
+    const dif_hmac_t *hmac, dif_hmac_endianness_t endianness) {
+  if (hmac == NULL) {
+    return kDifHmacBadArg;
+  }
+
+  switch (endianness) {
+    case kDifHmacEndiannessBig:
+      mmio_region_nonatomic_set_bit32(hmac->base_addr, HMAC_CFG_REG_OFFSET,
+                                      HMAC_CFG_DIGEST_SWAP);
+      break;
+    case kDifHmacEndiannessLittle:
+      mmio_region_nonatomic_clear_bit32(hmac->base_addr, HMAC_CFG_REG_OFFSET,
+                                        HMAC_CFG_DIGEST_SWAP);
+      break;
+    default:
+      return kDifHmacError;
+  }
+
+  return kDifHmacOk;
+}
+
+dif_hmac_result_t dif_hmac_key_set(const dif_hmac_t *hmac, const uint8_t *key) {
+  if (hmac == NULL || key == NULL) {
+    return kDifHmacBadArg;
+  }
+
+  load_key(hmac, key);
+
+  return kDifHmacOk;
+}
+
+dif_hmac_fifo_result_t dif_hmac_fifo_push(const dif_hmac_t *hmac,
+                                          const void *data, size_t len,
+                                          size_t *bytes_sent) {
+  if (hmac == NULL || data == NULL) {
+    return kDifHmacFifoBadArg;
+  }
+
+  const uint8_t *data_sent = (const uint8_t *)data;
+
+  // Calculate misalignment, taking care to not exceed the value of |len|.
+  int unaligned_bytes;
+  if (len < sizeof(uint32_t)) {
+    unaligned_bytes = (int)len;
+  } else {
+    unaligned_bytes = (uintptr_t)data_sent % sizeof(uint32_t);
+  }
+
+  int words_remaining = (len - unaligned_bytes) / sizeof(uint32_t);
+  int bytes_remaining = (len - unaligned_bytes) % sizeof(uint32_t);
+
+  // Bring |data_sent| into alignment so we can use |read32()| in the next loop.
+  while (unaligned_bytes > 0 && get_fifo_available_space(hmac) > 0) {
+    mmio_region_write8(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, *data_sent);
+    data_sent += sizeof(uint8_t);
+    --unaligned_bytes;
+  }
+
+  // Write whole words until the FIFO fills or there are no words left to send.
+  while (words_remaining > 0 && get_fifo_available_space(hmac) > 0) {
+    uint32_t word = read_32(data_sent);
+
+    mmio_region_write32(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, word);
+    data_sent += sizeof(uint32_t);
+    --words_remaining;
+  }
+
+  // Once all words are written write remaining bytes individually.
+  while (words_remaining == 0 && bytes_remaining > 0 &&
+         get_fifo_available_space(hmac) > 0) {
+    mmio_region_write8(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, *data_sent);
+    data_sent += sizeof(uint8_t);
+    --bytes_remaining;
+  }
+
+  if (bytes_sent != NULL) {
+    *bytes_sent = len - (words_remaining * sizeof(uint32_t)) - bytes_remaining;
```

Comment:
Is this calculation correct? what if `get_fifo_available_space(hmac) == 0` when `unaligned_bytes` is not yet zero?

Reply:
Another great catch, thanks!